### PR TITLE
fix: fix UAT tests to make it work on GKE COS environment

### PR DESCRIPTION
**Pull request by deesharma24**
## Summary
UAT Tests are failing in GKE env because the test is not able to fund gpu driver pod in gpu-operator namespace.

1. use the dcgm pod to inject kernel error related to xid79,xid119.
2. write the error to the /dev/kmsg because logger is not available.
3.  Added a initial test to verify whether driver pod is available or not in gpu-operator/kube-system namespace.

Unit test:-
1. Tested the script on the impacted cluster and able to inject and see the error and the node got quarantined.
```

  - lastHeartbeatTime: "2026-03-10T08:00:30Z"
    lastTransitionTime: "2026-03-10T06:54:15Z"
    message: 'ErrorCode:79 PCI:0002:00:00 MESSAGE=[9999999.200000] NVRM: Xid (PCI:0002:00:00):
      79, pid=1582259, name=nvc:[driver], GPU has fallen off the bus. Recommended
      Action=RESTART_BM;ErrorCode:79 PCI:0002:00:00 MESSAGE=[6085126.134786] NVRM:
      Xid (PCI:0002:00:00): 79, pid=1582259, name=nvc:[driver], GPU has fallen off
      the bus. Recommended Action=RESTART_BM;'
    reason: SysLogsXIDErrorIsNotHealthy
    status: "True"
    type: SysLogsXIDError

```

## Type of Change
- [x] 🐛 Bug fix
- [ ] ✨ New feature
- [ ] 💥 Breaking change
- [ ] 📚 Documentation
- [ ] 🔧 Refactoring
- [ ] 🔨 Build/CI

## Component(s) Affected
- [ ] Core Services
- [ ] Documentation/CI
- [ ] Fault Management
- [ ] Health Monitors
- [ ] Janitor
- [x] Other: ____________

## Testing
- [x] Tests pass locally
- [ ] Manual testing completed
- [x] No breaking changes (or documented)

## Checklist
- [x] Self-review completed
- [ ] Documentation updated (if needed)
- [x] Ready for review

## Summary by CodeRabbit

* **Tests**
  * Enhanced GPU driver verification to ensure a driver pod is present before GPU monitoring checks.
  * Improved GPU monitoring tests to better detect hardware errors (XID, SXID) and to inject and read messages via the monitoring agent context.
  * Updated test timeout handling with periodic progress logging for clearer observability during long-running checks.

### CODE REVIEW

File: tests/uat/tests.sh [hunk 1]
```diff
@@ -376,15 +398,15 @@ test_xid_monitoring_syslog() {
...
-    local driver_pod
-    driver_pod=$(kubectl get pods -n gpu-operator -l app=nvidia-driver-daemonset -o jsonpath="{.items[?(@.spec.nodeName=='$gpu_node')].metadata.name}" | head -1)
-
-    if [[ -z "$driver_pod" ]]; then
-        error "No driver pod found on node $gpu_node"
+    local dcgm_pod
+    dcgm_pod=$(kubectl get pods -n gpu-operator -l app=nvidia-dcgm \
+        -o jsonpath="{.items[?(@.spec.nodeName=='$gpu_node')].metadata.name}" | head -1)
+    if [[ -z "$dcgm_pod" ]]; then
+        error "No DCGM pod found on node $gpu_node"
```
[the remark this answers is not in the document]
Comment (deesharma24) [reply tests/uat/tests.sh:405]: Actually, it a pod is bad , then the test will fail which we want.So, there should be no issue with this.

File: tests/uat/tests.sh [hunk 2]
```diff
@@ -98,6 +98,26 @@ get_gpu_node_with_healthy_syslog_monitor() {
...
+        return
+    fi
+    pod=$(kubectl get pods -n kube-system -l k8s-app=nvidia-driver-installer --field-selector=status.phase=Running \
+        -o jsonpath="{.items[?(@.spec.nodeName=='$node')].metadata.name}" 2>/dev/null | head -1)
+    if [[ -n "$pod" ]]; then
+        log "Driver pod running: kube-system/$pod"
+        return
+    fi
+    error "No running driver pod found on node $node (checked gpu-operator and kube-system namespaces)"
+}
```
[the remark this answers is not in the document]
Comment (deesharma24) [reply tests/uat/tests.sh:118]: - If kubectl returns 0 but empty output → pod is empty → falls through to kube-system check. This is the normal "no driver pod in this namespace" case.
  - If kubectl errors out (network, RBAC) → set -e kills the script → test fails. This is correct — if we can't talk to the API server, we shouldn't silently skip to the next namespace.

  The only problematic case would be if kubectl fails for the gpu-operator lookup due to a namespace-specific RBAC issue but kube-system would have worked. That's an edge case not worth handling — if RBAC is broken for one namespace, the cluster setup is wrong and the test should fail.

  So the current behavior is actually fine. No change needed.

File: tests/uat/tests.sh [hunk 3]
```diff
@@ -98,6 +98,26 @@ get_gpu_node_with_healthy_syslog_monitor() {
...
+        log "Driver pod running: gpu-operator/$pod"
+        return
+    fi
+    pod=$(kubectl get pods -n kube-system -l k8s-app=nvidia-driver-installer --field-selector=status.phase=Running \
+        -o jsonpath="{.items[?(@.spec.nodeName=='$node')].metadata.name}" 2>/dev/null | head -1)
+    if [[ -n "$pod" ]]; then
+        log "Driver pod running: kube-system/$pod"
+        return
+    fi
+    error "No running driver pod found on node $node (checked gpu-operator and kube-system namespaces)"
```
Comment (lalitadithya) [inline tests/uat/tests.sh]: Can we make this a warn since in some envs the driver may be preinstalled on the host and in such envs this will cause the test to fail

Comment (deesharma24) [reply tests/uat/tests.sh]: done.

**Review by lalitadithya — APPROVED** [review]
